### PR TITLE
Enable Roslyn analyzers in build process

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

- Enable Roslyn analyzers in build process
- Force rule IDE0044: Add readonly modifier
- Add commit to fix the rule IDE0044

## PR Context

Analyzers help to make code high quality.
See .Net Runtime experience https://github.com/dotnet/runtime/issues/30740

We should benefit from C# analyzers too.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/powershell/11916)

### CODE REVIEW

File: PowerShell.sln [hunk 1]
```diff
@@ -50,52 +45,52 @@ Global
 		Release|Any CPU = Release|Any CPU
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
-		{8359D422-E0C4-4A0D-94EB-3C9DD16B7932}.CodeCoverage|Any CPU.ActiveCfg = CodeCoverage|Any CPU
```
Comment (TravisEz13) [inline PowerShell.sln]: Please revert this changes

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
too big to review, PR review tool doesn't function

**Comment by TravisEz13** [comment]
Perhaps you can enable one rule at a time to reduce the file changes.  I cannot jump to files due to the large number of changes.

File: PowerShell.sln (hunk 1 above)
Comment (iSazonov) [reply PowerShell.sln]: Only the line or all such lines?

**Comment by iSazonov** [comment]
@TravisEz13 You can review commit by commit. Yes, only one rule is forced in the PR. This add "readonly" qualifier. I will split the large commit by 20-30 files.

File: PowerShell.sln (hunk 1 above)
Comment (iSazonov) [reply PowerShell.sln]: Reverted all.

File: .editorconfig [hunk 2]
```diff
@@ -135,6 +135,21 @@ csharp_style_inlined_variable_declaration = true:suggestion
 csharp_style_throw_expression = true:suggestion
```
Comment (PoshChan) [inline .editorconfig]: @iSazonov, your last commit had 2 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/docs/cmdlet-example/command-line-simple-example.md.https://www.microsoft.com/net/core#linuxubuntu should work
```powershell
Failed to complete request to "https://www.microsoft.com/net/core#linuxubuntu". Response status code does not indicate success: 400 (Bad Request).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/docs/building/windows-core.md.https://www.microsoft.com/net/core#windowscmd should work
```powershell
Failed to complete request to "https://www.microsoft.com/net/core#windowscmd". Response status code does not indicate success: 400 (Bad Request).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```

**Comment by TravisEz13** [comment]
Git hub doesn't have a mechanism for tracking reviews, commit by commit.  I can see if reviewable still works.

**Comment by TravisEz13** [comment]
@PoshChan Please remind me in 2 days

**Comment by PoshChan** [comment]
@TravisEz13, this is the reminder you requested 2 days ago

**Comment by iSazonov** [comment]
@daxian-dbw @TravisEz13 @adityapatwardhan The PR changes many files. Please review to avoid merge conflicts.

See https://github.com/dotnet/runtime/issues/30740 for context.

File: .editorconfig (hunk 2 above)
Comment (PoshChan) [inline .editorconfig]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
Test-Connection.TraceRoute.TraceRoute works
```powershell
Expected exactly '10.79.1.68', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.165.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 258
258:             $result[0].TargetAddress | Should -BeExactly $realAddress
```

**Comment by iSazonov** [comment]
@TravisEz13 Could you please update your review?

**Comment by iSazonov** [comment]
@xtqqczze The PR overlap your PR. Could you please review?

**Review by TravisEz13 — COMMENTED** [review]
> Reviewed 31 of 366 files at r1.
> *[Reviewable](https://reviewable.io/reviews/powershell/powershell/11916)* status: 31 of 366 files reviewed (waiting on @adityapatwardhan, @anmenaga, @daxian-dbw, @iSazonov, and @SteveL-MSFT)

**Comment by TravisEz13** [comment]
I'll try to review a portion each day.

**Comment by TravisEz13** [comment]
@PoshChan  Please remind me in 1 day.

**Comment by PoshChan** [comment]
@TravisEz13, this is the reminder you requested 1 day. ago

**Comment by xtqqczze** [comment]
This PR does not seem to enable any analyzers during the build process. If it did then the build would fail since `IDE1006.severity` is set to `error` and there are 904 IDE1006 rule violations.

**Comment by xtqqczze** [comment]
Currently IDE code style rules do not apply during builds, see https://github.com/dotnet/roslyn/issues/33558

**Comment by iSazonov** [comment]
@xtqqczze Yes, you are right about that IDE rules. The PR is only initial for the enhancement of our build process.  Later we can easily add other rules and force then as needed. Also I hope the .Net issue will fixed in some way.

File: src/System.Management.Automation/security/nativeMethods.cs [hunk 3]
```diff
@@ -1081,36 +1081,36 @@ internal static DWORD DestroyWintrustDataStruct(WINTRUST_DATA wtd)
...
+            private readonly BOOL _fTestCert;
+            private readonly DWORD _dwRevokedReason;
+            private readonly DWORD _dwConfidence;
+            private readonly DWORD _dwError;
+            private readonly IntPtr _pTrustListContext; // CTL_CONTEXT*
+            private readonly BOOL _fTrustListSignerCert;
+            private readonly IntPtr _pCtlContext; // PCCTL_CONTEXT
+            private readonly DWORD _dwCtlError;
+            private readonly BOOL _fIsCyclic;
+            private readonly IntPtr _pChainElement; // PCERT_CHAIN_ELEMENT
```
Comment (vexx32) [inline src/System.Management.Automation/security/nativeMethods.cs:1098]: We don't provide a constructor for these and they're readonly. How would these be set?

Comment (vexx32) [reply src/System.Management.Automation/security/nativeMethods.cs:1098]: Same question for the rest of the structs in this file actually.

**Review by vexx32 — APPROVED** [review]
Overall these changes look good. I have a couple questions/comments on a few of them. 🙂

File: src/Microsoft.Management.Infrastructure.CimCmdlets/CimCmdletModuleInitialize.cs [hunk 4]
```diff
@@ -79,21 +79,21 @@ internal CimCmdletAliasEntry(string name, string value)
...
-            private string _value = string.Empty;
+            private readonly string _value = string.Empty;
 
             /// <summary>
             /// The string defining real cmdlet name.
             /// </summary>
             internal ScopedItemOptions Options { get { return this._options; } }
 
-            private ScopedItemOptions _options = ScopedItemOptions.AllScope | ScopedItemOptions.ReadOnly;
+            private readonly ScopedItemOptions _options = ScopedItemOptions.AllScope | ScopedItemOptions.ReadOnly;
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/Microsoft.Management.Infrastructure.CimCmdlets/CimCmdletModuleInitialize.cs:96]: It is still not merged.

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 5]
```diff
@@ -4100,7 +4100,7 @@ internal void ImportCmdletsFromAssembly(Assembly assembly, PSModuleInfo module)
         /// <summary>
         /// This is the default function to use for tab expansion.
         /// </summary>
-        private static string s_tabExpansionFunctionText = @"
+        private static readonly string s_tabExpansionFunctionText = @"
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4103]: It is static. If it is wrong we should fix this in another PR.

File: src/System.Management.Automation/engine/hostifaces/HostUtilities.cs [hunk 6]
```diff
@@ -58,21 +58,21 @@ public static class HostUtilities
             $foundSuggestion
         ";
 
-        private static string s_createCommandExistsInCurrentDirectoryScript = @"
+        private static readonly string s_createCommandExistsInCurrentDirectoryScript = @"
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/HostUtilities.cs:61]: The same.

File: src/System.Management.Automation/security/nativeMethods.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/security/nativeMethods.cs:1098]: I'd ask why they private :-)

Looks like false positives.

Comment (vexx32) [reply src/System.Management.Automation/security/nativeMethods.cs:1098]: Yeah both are good questions here 😁 

Not sure how this struct is actually used to be honest, but I don't see how it would be useful at the moment.

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4103]: Yeah that's why I asked. Maybe make a note of these and we can have another look in a later PR, see if it's worth making them `const`. 🙂

File: src/System.Management.Automation/security/nativeMethods.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/security/nativeMethods.cs:1098]: In fact, all these structures are created in P/Invoke, they are really readonly for C# and "private" here is "named unused".

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/InitialSessionState.cs:4103]: It is used only once. I believe the const would be saved in a static area too and it makes no sense to change - we get no perf benefits.

**Comment by iSazonov** [comment]
@TravisEz13 Could you please continue?

**Comment by xtqqczze** [comment]
@iSazonov something to be aware of: https://github.com/dotnet/roslyn/issues/43051

**Comment by iSazonov** [comment]
@xtqqczze Thanks! We use `warnaserror` in the project.

**Comment by iSazonov** [comment]
@TravisEz13 Could you please continue the code review?

**Review by TravisEz13 — COMMENTED** [review]
> Reviewed 35 of 366 files at r1.
> *[Reviewable](https://reviewable.io/reviews/powershell/powershell/11916)* status: 57 of 366 files reviewed (waiting on @adityapatwardhan, @anmenaga, @daxian-dbw, @iSazonov, @SteveL-MSFT, @TravisEz13, and @vexx32)

**Comment by xtqqczze** [comment]
@iSazonov It seems this PR is stale, we could continue with IDE0044 in #13880? We could split the changes into multiple PRs for ease of review.

**Comment by iSazonov** [comment]
> @iSazonov It seems this PR is stale, we could continue with IDE0044 in #13880? We could split the changes into multiple PRs for ease of review.

Yes.
